### DISCUSSION TITLE
Resolved: What should go into the next release?

### DISCUSSION THREAD

**Discussion by Kixunil** [rust-bitcoin/rust-bitcoin#1501] [General]
EDIT: This refers to the current released version: https://crates.io/crates/bitcoin/0.30.0

Draft list:

- [x] everything in `master`, obviously
- [x] All issues currently marked for 0.30 milestone
- [x] All API holes
- [x] removal of `hashes::hex`
- [x] Various taproot improvements (recent issues)
- [x] `PushBytes` unsized type
- [ ] complete polishing of error handling (remove error handling epic from the milestone)
- [ ] `Script` tagging

Unchecked box means "we want to do it eventually but probably skip the next release"

One I'm not sure about is `PushBytes`. It removes some panics but it's also somewhat big and depends on `hex_lit` among other things. Also if people can still screw `Script` in different ways (bad opcodes) I'm not sure if it's that useful without the remaining improvements.

Feel free to suggest other items or argue for/against.

**Comment by apoelstra** [comment]
Looks good! I'd also add the following crate smashing items (I'm not sure how complete they are):

-[ ] get rid of `util` module (I think this is done?)
-[ ] introduce `crypto` module (this is just one PR that I think we're close on)

I'd also like to get the `string` crate in, with hex sorted out, but I wouldn't hold up the release for it.

**Reply by tcharding** [reply]
introduce `crytpo` module? Did you mean crate?

**Reply by tcharding** [reply]
Yep, util is done - the only thing there is deprecated code that we can get rid of in 0.31.0

**Reply by apoelstra** [reply]
Yeah, I meant crypto crate, sorry.

**Reply by tcharding** [reply]
There is no open PR for making the `crypto` module into a crate. I had a quick go and there is still some other crates smashing things that need doing first, shown here by the use statements in the `crypto` module
```
gg 'use crate::'                                                                                                                                                    ✘ 1 
ecdsa.rs:14:use crate::prelude::*;
ecdsa.rs:15:use crate::hashes::hex::{self, FromHex};
ecdsa.rs:16:use crate::sighash::{EcdsaSighashType, NonStandardSighashType};
key.rs:9:use crate::prelude::*;
key.rs:17:use crate::{base58, io};
key.rs:18:use crate::network::constants::Network;
key.rs:19:use crate::hashes::{Hash, hash160, hex, hex::FromHex};
key.rs:20:use crate::hash_types::{PubkeyHash, WPubkeyHash};
key.rs:547:    use crate::io;
key.rs:551:    use crate::hashes::hex::{FromHex, ToHex};
key.rs:552:    use crate::network::constants::Network::Testnet;
key.rs:553:    use crate::network::constants::Network::Bitcoin;
key.rs:554:    use crate::address::Address;
schnorr.rs:16:use crate::prelude::*;
schnorr.rs:18:use crate::taproot::{TapBranchHash, TapTweakHash};
schnorr.rs:19:use crate::sighash::SchnorrSighashType;
```

I rekon this stuff will need discussing a bunch and as such we probably should not be aiming to make a `crypto` crate as part of 0.30

**Reply by apoelstra** [reply]
Ok, thanks for checking! Agreed, let's not do `crypto` for this release.

**Comment by tcharding** [comment]
I had a look through all issues back to December and fixed a few, are there more @Kixunil that you were referring to with "Various taproot improvements (recent issues)"?

**Comment by tcharding** [comment]
I've converted any of my PRs that are not for this release to draft. Please consider reviewing everything I have open with the caveat that #1573 should be the last to be considered.

**Reply by sanket1729** [reply]
Sorry for being inactive, I should be active again after Feb 15th. Will try my best to offer reviews for open PRs for this release. Unfortunately, would not have time to work on any of my open PRs.

**Reply by tcharding** [reply]
No worries man, cheers.

**Comment by tcharding** [comment]
The upcoming release is going to require releases of the whole stack. I had a quick squiz at secp256k1 and it looks ready (excluding changes already seen in the hashes work). If anything needs doing there please poke me.

**Comment by Kixunil** [comment]
The milestone says only formatting is missing. I'm not sure this is accurate. Please set the milestone for things you want to get in.

**Reply by apoelstra** [reply]
I think https://github.com/rust-bitcoin/rust-bitcoin/pull/1577 maybe should go in? cc @tcharding 

Let's also get https://github.com/rust-bitcoin/rust-bitcoin/pull/1651 in (I'll run local CI on this within the hour and ACK/merge it. Looks good to me.)

**Reply by Kixunil** [reply]
Also #1597 would be nice.

**Reply by tcharding** [reply]
I have 5 open PRs that I'd like to be considered please (PR filter: `is:open is:pr author:tcharding -is:draft`). I've added all of them _except_ the hex crate one to milestone 0.30 so they can be found that way as well. I was hoping to discuss the hex crate one after every other PR for this release (of mine and others) is merged.

**Reply by tcharding** [reply]
Looks to me like I'm the only one with open PRs left for this release, that means the formatting one can go in now if acked please.

https://github.com/rust-bitcoin/rust-bitcoin/pull/1111

**Reply by sanket1729** [reply]
https://github.com/rust-bitcoin/rust-bitcoin/pull/1533 has some bug fixes that should go in

**Comment by Kixunil** [comment]
FYI I tried updating one of my larger projects and it went mostly fine, opened issues for annoying parts.

**Reply by tcharding** [reply]
Ah sweet, I was hoping that miniscript was an unusual user and others would be better.

**Comment by tcharding** [comment]
I've added "release notes metion" label to PRs that seem like candidates to go into the upcoming release changelog, please add/remove the label to PRs you'd like to see, or not see, mentioned.

**Comment by tcharding** [comment]
Now stale, closing.
